sdk/core/azure-core/src/main/java/com/azure/core/util/SharedExecutorService.java

```diff
@@ -0,0 +1,352 @@
+        return ensureNotShutdown().invokeAny(tasks);
+    }
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+        throws InterruptedException, ExecutionException, TimeoutException {
+        return ensureNotShutdown().invokeAny(tasks, timeout, unit);
+    }
+
+    private ExecutorService ensureNotShutdown() {
+        return wrappedExecutorService.updateAndGet(wrapper -> {
+            if (wrapper == null || wrapper.executorService.isShutdown() || wrapper.executorService.isTerminated()) {
+                return createSharedExecutor();
+            } else {
+                return wrapper;
+            }
+        }).executorService;
+    }
+
+    private static ExecutorWithMetadata createSharedExecutor() {
+        ThreadFactory threadFactory;
+        if (VIRTUAL_THREAD_SUPPORTED && THREAD_POOL_VIRTUAL) {
+            try {
+                LOGGER.verbose("Attempting to create a virtual thread factory.");
+                threadFactory = createVirtualThreadFactory();
+                LOGGER.verbose("Successfully created a virtual thread factory.");
+            } catch (Exception e) {
+                LOGGER.info("Failed to create a virtual thread factory, falling back to non-virtual threads.", e);
+                threadFactory = createNonVirtualThreadFactory();
+            }
+        } else {
+            threadFactory = createNonVirtualThreadFactory();
+        }
+
+        ExecutorService executorService = new ThreadPoolExecutor(0, THREAD_POOL_SIZE, THREAD_POOL_KEEP_ALIVE_MILLIS,
+            TimeUnit.MILLISECONDS, new SynchronousQueue<>(), threadFactory);
+        Thread shutdownThread = CoreUtils.createExecutorServiceShutdownThread(executorService, Duration.ofSeconds(5));
+        CoreUtils.addShutdownHookSafely(shutdownThread);
+
+        return new ExecutorWithMetadata(executorService, shutdownThread);
+    }
+
+    private static ThreadFactory createVirtualThreadFactory() throws Exception {
+        Object virtualThreadBuilder = GET_VIRTUAL_THREAD_BUILDER.invokeStatic();
+        SET_VIRTUAL_THREAD_BUILDER_THREAD_NAME.invokeWithArguments(virtualThreadBuilder, AZURE_SDK_THREAD_NAME);
+        return (ThreadFactory) CREATE_VIRTUAL_THREAD_FACTORY.invokeWithArguments(virtualThreadBuilder);
+    }
+
+    private static ThreadFactory createNonVirtualThreadFactory() {
+        return r -> {
+            Thread thread = new Thread(r, AZURE_SDK_THREAD_NAME + AZURE_SDK_THREAD_COUNTER.getAndIncrement());
+            thread.setDaemon(true);
+
+            return thread;
+        };
+    }
+
+    private static final class ExecutorWithMetadata {
+        private final Thread shutdownThread;
+        private final ExecutorService executorService;
+
+        ExecutorWithMetadata(ExecutorService executorService, Thread shutdownThread) {
+            this.executorService = executorService;
+            this.shutdownThread = shutdownThread;
+        }
+
+        void shutdown() {
```

Comment:
Since we don't allow setting the executor service (that is commented out), this method is not required as we don't support shutting the shared executor service down directly. It will always be done through the shutdown hook.

Reply:
I'll comment this out as well and use ExecutorService directly given the recent design change